### PR TITLE
refactor(state-table): Avoid exposing vnode-prefixed key in state table iter API

### LINKED ISSUE #11869: using of `risingwave_storage::table::merge_sort` may be incorrect due to vnode prefix returned by table iter stream
Due to https://github.com/risingwavelabs/risingwave/issues/11673, the following calls to `merge_sort` may lead to incorrect result:

- https://github.com/risingwavelabs/risingwave/blob/451c3873ff2fc713fca3214e88aba94e3154ce61/src/stream/src/common/table/state_table.rs#L894
- https://github.com/risingwavelabs/risingwave/blob/451c3873ff2fc713fca3214e88aba94e3154ce61/src/stream/src/executor/sort_buffer.rs#L219

This issue also occurred when implementing OverWindow range cache, I made a workaround by wrapping the iter stream: https://github.com/risingwavelabs/risingwave/blob/b0f1cafc31f593132c633546eed02e602688acaf/src/stream/src/executor/over_window/over_partition.rs#L519-L550

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix #11869 #11673

- `KeyedRow` struct is introduced and is used as the returning item type for storage table and state table public iter APIs.
- State table won't have separate APIs for `iter_row_xxx` and `iter_row_and_key_xxx`. Caller should decide whether it needs the key part of the `KeyedRow` and can choose to discard that by calling `KeyedRow::into_row`
- `merge_sort` is now based on `KeyedRow`
- `key_part` and `vnode_part` methods are added to `TableKey` to retrieve the corresponding parts of the key.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/common/table/state_table.rs
Comment (kwannoel) [inline src/stream/src/common/table/state_table.rs:1140]: Should we add a comment here to warn users that `raw key-values` means encoded bytes of `vnode` will be a prefix of the key?

Perhaps we should also add a new interface (can be a separate PR) that iterates over `pk` and `row`, without `vnode`, so we can use it in `merge_sort`.

Otherwise correctness of `merge_sort` is affected, because as mentioned here https://github.com/risingwavelabs/risingwave/issues/11869, we will be sorting on the vnode prefix too.

File: src/stream/src/common/table/state_table.rs
Comment (kwannoel) [inline src/stream/src/common/table/state_table.rs:887]: I will fix this in a separate PR, I think it should iter on pk here, rather than raw key.

Otherwise https://github.com/risingwavelabs/risingwave/issues/11869 is encountered.

File: src/stream/src/common/table/state_table.rs
Comment (BugenZhao) [reply src/stream/src/common/table/state_table.rs:1140]: I guess we can prevent developers from misusing these interfaces once and for all with the help of the type system. Like `FullKey<Bytes>`, we can also have `TableKey` or `RealKey` as the return types. We may even unify all these interfaces with the generic of `impl FromRawKv`.

```rust
pub trait FromRawKv {
  fn from_raw(key: FullKey<Bytes>, value: Bytes) -> Self;
}

impl FromRawKv for OwnedRow { .. }
impl FromRawKv for (TableKey, OwnedRow) { .. }
impl FromRawKv for (RealKey, OwnedRow) { .. }

impl StateTable {
  fn iter<R: FromRawKv>(..) -> impl TryStream<R> { .. }
}
```

File: src/storage/src/table/merge_sort.rs
Comment (stdrc) [inline src/storage/src/table/merge_sort.rs:29]: The `K` here now will be `(VirtualNode, Bytes)` but I think what we need is the `Bytes` part.

File: src/stream/src/common/table/state_table.rs
Comment (hzxa21) [reply src/stream/src/common/table/state_table.rs:1140]: > Otherwise correctness of merge_sort is affected, because as mentioned here https://github.com/risingwavelabs/risingwave/issues/11869, we will be sorting on the vnode prefix too.

I thought the merge_sort is only used to merge data from different vnodes. Anyway, now that the interface returns a tuple (VirtualNode, Bytes) instead of raw Bytes as the key so I think it is clear enough. If one wants to merge across vnodes, the returning stream can be used directly. If one wants to merge by ignoring vnodes, the returning stream key can be mapped to ignore the VirtualNode part of the key before doing merge_sort.

Comment (hzxa21) [reply src/stream/src/common/table/state_table.rs:1140]: > I guess we can prevent developers from misusing these interfaces once and for all with the help of the type system. Like `FullKey<Bytes>`, we can also have `TableKey` or `RealKey` as the return types. We may even unify all these interfaces with the generic of `impl FromRawKv`.
> 
> ```rust
> pub trait FromRawKv {
>   fn from_raw(key: FullKey<Bytes>, value: Bytes) -> Self;
> }
> 
> impl FromRawKv for OwnedRow { .. }
> impl FromRawKv for (TableKey, OwnedRow) { .. }
> impl FromRawKv for (RealKey, OwnedRow) { .. }
> 
> impl StateTable {
>   fn iter<R: FromRawKv>(..) -> impl TryStream<R> { .. }
> }
> ```

Good idea. Let me try.

File: src/stream/src/common/table/state_table.rs
Comment (hzxa21) [reply src/stream/src/common/table/state_table.rs:887]: I guess this is a bug prior to this PR? I thought the intention is to merge_sort with vnode included. This is another bug caused by the confusing API lol

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: > I will fix this in a separate PR, I think it should iter on pk here, rather than raw key.

Couldn't quite get it here. I think this by splitting vnode and pk bytes, the `merge_sort` can be corrected, and then #11869 will no longer be an issue.

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: > I thought the intention is to merge_sort with vnode included.

No, `merge_sort` is intended to sort rows from different vnodes by their pk without vnode prefix.

Comment (kwannoel) [reply src/stream/src/common/table/state_table.rs:887]: > I guess this is a bug prior to this PR? I thought the intention is to merge_sort with vnode included. This is another bug caused by the confusing API lol

Yes it is prior to this PR, out of scope and can be fixed separately.

File: src/storage/src/table/merge_sort.rs
Comment (hzxa21) [reply src/storage/src/table/merge_sort.rs:29]: Since `VirtualNode` is passed to `iter_xx_with_pk_range` AP, I thought it will be helpful to expose vnode as part of the key to the upper layer as well and upper layer can decide whether they want to discard it or not.

Comment (hzxa21) [reply src/storage/src/table/merge_sort.rs:29]: If we can confirm that vnode is not needed in the returning key, I can change that as well.

File: src/stream/src/common/table/state_table.rs
Comment (kwannoel) [reply src/stream/src/common/table/state_table.rs:887]: > Couldn't quite get it here. I think by splitting vnode and pk bytes in this PR, the merge_sort can be corrected, and then https://github.com/risingwavelabs/risingwave/issues/11869 will no longer be an issue.

I mean here we still use `iter_kv_with_pk_range`. Underlying that is:
```
pub type StateStoreIterItem = (FullKey<Bytes>, Bytes);
```

The `FullKey` includes `vnode` as well. So if you look later in the code, that still uses `deserialize_pk_with_vnode` to deserialize, which is incorrect.

After this PR, `iter_row_and_key_with_pk_range` will return :
```
pub type RowStreamWithKey<'a, S: StateStore, SD: ValueRowSerde + 'a> =
    impl Stream<Item = StreamExecutorResult<((VirtualNode, Bytes), OwnedRow)>> + 'a;
```

We want the new iter to be something like `iter_pk_and_row_with_pk_range`.

Which is just a simple wrapper around `iter_row_and_key_with_pk_range`. This will drop the vnode, part, and ensure when `merge_sort` happens, it will only sort by `pk`.

File: src/storage/src/table/merge_sort.rs
Comment (kwannoel) [reply src/storage/src/table/merge_sort.rs:29]: I prefer to keep existing api in this PR as is for flexibility. Some components will still want to iter with vnode, e.g. arrangement backfill.

Introducing a new iter which just drops the vnode part will be good enough IMO.

File: src/stream/src/common/table/state_table.rs
Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: Then we're at the same page here. While from my perspective, it doesn't need to be made too complex.

Since we provide `vnode` and `pk_range` separately when calling `iter_kv_with_pk_range`, it should return `vnode` and `pk` separately, I guess that why this PR exists. And once we have separated `vnode` and `pk`, we need to rethink `merge_sort`. AFAIK from its very first use in streaming executors, it's supposed to sort rows from different iter streams by `pk` instead of `vnode + pk`.

So the two issues are highly related and `merge_sort` can be fixed in this PR by just use `peeked = stream.next().await?.0.1`, or, for better readability we may consider put `vnode` and `pk` in a struct instead of anonymous tuple.

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: My point is that we don't need introduce a new state table method or wrap the streams before calling `merge_sort` to fix it.

Comment (hzxa21) [reply src/stream/src/common/table/state_table.rs:887]: >  for better readability we may consider put vnode and pk in a struct instead of anonymous tuple.


> My point is that we don't need introduce a new state table method or wrap the streams before calling merge_sort to fix it.

+1 for both

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: Oh I realized that we can't get `1` from the key in `merge_sort`. Just ignore me😇

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: Another reasonable solution may be wrapping `vnode` and `pk` in  a struct and impl `Ord` for it to make it sort by `pk`. IMO it's reasonable because `pk` is the key the caller want to care about, state table should hide vnode concept (for almost all executors, the order of vnodes doesn't matter).

> Some components will still want to iter with vnode, e.g. arrangement backfill.

For some of such executors that need to care about vnode, they may simply chain all streams when calling `iter_vnodes` and creating iteration streams.

Comment (stdrc) [reply src/stream/src/common/table/state_table.rs:887]: Anyway, just some concern on elegance, no strong opinion. All solutions mentioned in previous comments LGTM🤪

**Comment by hzxa21** [comment]
Update on the new commit:
- `KeyedRow` struct is introduced and is used as the returning item type for storage table and state table public iter APIs.
- State table won't have separate APIs for `iter_row_xxx` and `iter_row_and_key_xxx`. Caller should decide whether it needs the key part of the `KeyedRow` and can choose to discard that by calling `KeyedRow::into_row`
- `merge_sort` is now based on `KeyedRow`
- `key_part` and `vnode_part` methods are added to `TableKey` to retrieve the corresponding parts of the key.

File: src/storage/src/table/mod.rs
Comment (stdrc) [inline src/storage/src/table/mod.rs:238]: I'm not quite sure but maybe it's a good idea to `impl Row for KeyedRow`? `Row` trait has a method called `into_owned_row`.

File: src/storage/src/table/merge_sort.rs
Comment (stdrc) [inline src/storage/src/table/merge_sort.rs:24]: This seems can be removed.

File: src/storage/src/row_serde/row_serde_util.rs
Comment (stdrc) [inline src/storage/src/row_serde/row_serde_util.rs:42]: Maybe we can use `let key = TableKey(key)` `key.vnode_part()` `key.key_part()` here.

File: src/stream/src/common/table/state_table.rs
Comment (stdrc) [inline src/stream/src/common/table/state_table.rs:896]: I guess `deserialize_pk_with_vnode` call here is not needed any more.

**Review by stdrc — APPROVED** [review]
Looks clean to me! #11869 can be closed as completed once this PR get merged.

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM!

File: src/stream/src/common/table/state_table.rs
Comment (hzxa21) [reply src/stream/src/common/table/state_table.rs]: Good catch.

File: src/storage/src/row_serde/row_serde_util.rs
Comment (hzxa21) [reply src/storage/src/row_serde/row_serde_util.rs:42]: This requires a bigger refactoring since the memtable key is in `Bytes` not `TableKey` type. Given that the memtable is already in storage, I will try to expose the `TableKey` struct in memtable as well in a separate PR and fix this issue then.

File: src/storage/src/table/mod.rs
Comment (hzxa21) [reply src/storage/src/table/mod.rs]: After checking all the usage, I think `impl Deref` is more appropriate here.

**Comment by hzxa21** [comment]
All comments are addressed. @stdrc @BugenZhao PTAL.

**Review by stdrc — APPROVED** [review]
LGTM!

**Pull request merged**
